In [1]:
'''
MSSV: 21110429
Tên: Nguyễn Quang Trường
Bài Thực Hành Tuần 4
Lớp: 20TH_KDL
'''

'\nMSSV: 21110429\nTên: Nguyễn Quang Trường\nBài Thực Hành Tuần 4\nLớp: 20TH_KDL\n'

# Bài 4: Khai Thác Mẫu Kết Hợp

In [2]:
pip install apyori

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5974 sha256=07202a0aa2b77cc9541d117fb1b00f25047b0fc6a958fd898f98e33933b774c5
Note: you may need to restart the kernel to use updated packages.  Stored in directory: c:\users\quang\appdata\local\pip\cache\wheels\32\2a\54\10c595515f385f3726642b10c60bf788029e8f3a1323e3913a
Successfully built apyori



In [3]:
from apyori import apriori
import pandas as pd
import numpy as np

In [4]:
data=pd.read_csv("data.csv",header=None)

In [5]:
data.head()

0      1      2       3     4      5
0  Wine  Chips  Bread  Butter  Milk  Apple
1  Wine    NaN  Bread  Butter  Milk    NaN
2   NaN    NaN  Bread  Butter  Milk    NaN
3   NaN  Chips    NaN     NaN   NaN  Apple
4  Wine  Chips  Bread  Butter  Milk  Apple

In [6]:
# chuyen DataFrame thanh dang danh sach cac list
transaction=[]
for i in range(0,data.shape[0]):
    transaction.append([str(data.values[i,j]) for j in range(0,data.shape[1])])

In [7]:
# Loai bo cac gia tri 'nan'
transaction_new=[]
for i in range(0, len(transaction)):
    transaction_new.append([])
    for j in transaction[i]:
        if j != 'nan':
            transaction_new[i].append(j)

In [8]:
transaction_new

[['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Bread', 'Butter', 'Milk'],
 ['Bread', 'Butter', 'Milk'],
 ['Chips', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Milk'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Apple'],
 ['Wine', 'Chips', 'Milk'],
 ['Wine', 'Bread', 'Apple'],
 ['Wine', 'Bread', 'Butter', 'Milk'],
 ['Chips', 'Bread', 'Butter', 'Apple'],
 ['Wine', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk'],
 ['Wine', 'Bread', 'Milk', 'Apple'],
 ['Wine', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Chips', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Bread', 'Butter', 'Milk', 'Apple'],
 ['Wine', 'Chips', 'Bread', 'Milk', 'Apple'],
 ['Chips']]

## Xây dựng theo hàm có sẵn trong python

### Xây dựng mô hình Apriori

In [9]:
assocciation_rules=apriori(transaction_new,min_support=0.5, min_confidence=0.7,min_lift=1.2)
assocciation_rules=list(assocciation_rules)

In [10]:
assocciation_rules

[RelationRecord(items=frozenset({'Bread', 'Milk', 'Butter'}), support=0.5, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Butter'}), items_add=frozenset({'Bread', 'Milk'}), confidence=0.7333333333333334, lift=1.241025641025641), OrderedStatistic(items_base=frozenset({'Bread', 'Milk'}), items_add=frozenset({'Butter'}), confidence=0.8461538461538461, lift=1.241025641025641)])]

Kết quả này là một tuple duy nhất trong danh sách các tuples đại diện cho các luật kết hợp phổ biến tìm thấy trong dữ liệu đầu vào bằng thuật toán Apriori.

* items trong tuple này là một frozenset chứa tất cả các đặc trưng được kết hợp với nhau, tức là "Butter", "Bread" và "Milk" được kết hợp với nhau.
* support cho biết tỷ lệ phần trăm của số lần xuất hiện của bộ các đặc trưng đó trong tất cả các giao dịch trong dữ liệu đầu vào, ở đây là 0.5, tức là bộ các đặc trưng này xuất hiện trong một nửa số lượng giao dịch.
* ordered_statistics là một danh sách các OrderedStatistic objects, mỗi object biểu diễn một cặp luật kết hợp theo quy tắc: "nếu bạn mua đặc trưng A, thì bạn cũng sẽ mua đặc trưng B".
    - Trong tuple này, có hai OrderedStatistic objects.
        + OrderedStatistic đầu tiên cho biết nếu người dùng mua "Butter", thì có khả năng hơn 73.33% rằng họ cũng sẽ mua "Bread" và "Milk" cùng lúc. Chỉ số lift của ordered statistic này là 1.241025641025641, cho thấy rằng mua "Butter" cùng với "Bread" và "Milk" có mức độ liên quan tương đối cao, so với một trường hợp không có sự kết hợp này.
        + OrderedStatistic thứ hai cho biết nếu người dùng mua cùng lúc "Bread" và "Milk", thì có khả năng hơn 84.61% rằng họ cũng sẽ mua "Butter" cùng lúc. Chỉ số lift của ordered statistic này cũng là 1.241025641025641, cho thấy mua "Bread" và "Milk" cùng với "Butter" cũng có mức độ liên quan tương đối cao.

### Xuất kết quả 

In [11]:
# Xuất kết quả

def output_solution(assocciation_rules):
    df = pd.DataFrame(columns=['Rule', 'Support', 'Confidence', 'Lift'])
    for r in assocciation_rules:
        lhs = ', '.join(list(r.ordered_statistics[0].items_base))
        rhs = ', '.join(list(r.ordered_statistics[0].items_add))
        support = r.support
        confidence = r.ordered_statistics[0].confidence
        lift = r.ordered_statistics[0].lift
        df = pd.concat([df, pd.DataFrame({'Rule': f"{lhs} => {rhs}", 'Support': support, 'Confidence': confidence, 'Lift': lift}, index=[0])], ignore_index=True)
        if len(r.ordered_statistics) > 1:
            for os in r.ordered_statistics[1:]:
                lhs = ', '.join(list(os.items_base))
                rhs = ', '.join(list(os.items_add))
                support = r.support
                confidence = os.confidence
                lift = os.lift
                df = pd.concat([df, pd.DataFrame({'Rule': f"{lhs} => {rhs}", 'Support': support, 'Confidence': confidence, 'Lift': lift}, index=[0])], ignore_index=True)

    print(df)

In [12]:
output_solution(assocciation_rules)

                    Rule Support Confidence      Lift
0  Butter => Bread, Milk     0.5   0.733333  1.241026
1  Bread, Milk => Butter     0.5   0.846154  1.241026


In [13]:
def inspect(output):
    lhs         = [tuple(result[2][0][0])[0] for result in output]
    rhs         = [tuple(result[2][0][1])[0] for result in output]
    support    = [result[1] for result in output]
    confidence = [result[2][0][2] for result in output]
    lift       = [result[2][0][3] for result in output]
    return list(zip(lhs, rhs, support, confidence, lift))


In [14]:
output_DataFrame = pd.DataFrame(inspect(assocciation_rules), columns = ['Left_Hand_Side', 'Right_Hand_Side', 'Support', 'Confidence', 'Lift'])


In [15]:
output_DataFrame

Left_Hand_Side Right_Hand_Side  Support  Confidence      Lift
0         Butter           Bread      0.5    0.733333  1.241026

## Xây dựng mô hình bằng hàm tự viết

### Các hàm hỗ trợ

In [16]:
def count_occurences(itemset, transactions): 
    """
    Feature: Number of occurrences of items in transactions
    Return: count items in transactions
    
    Arguments:
        itemset
        transactions
    """
    count=0
    for i in range(len(transactions)):
        if set(itemset).issubset(set(transactions[i])):
            count+=1 
    return count 

In [17]:
def get_frequent(itemsets,transactions, min_support, prev_discard):
   
    """
    Feature:find the frequent set in the transaction data set with a given min_support threshold value.
    Return:
        L: is the list of frequent episodes.
        supp_count: is the number of occurrences of the frequent itemsets in the transaction data set.
        new_discarded: is the list of unpopular episodes to be removed.
    Agrument:
        itemsets
        transactions
        min_suport
        prev_discard: contains a list of episodes that have 
                        been dropped in previous frequent set searches.
    """   
    L=[]
    supp_count=[]
    new_discarded=[]
    num_trans=len(transactions)
    k=len(prev_discard.keys())
    for s in range(len(itemsets)):
        discarded_before= False
        if k >0:
            for it in prev_discard[k]:
                if set(it).issubset(set(itemsets[s])):
                    discarded_before=True
                    break
        if not discarded_before:
            count = count_occurences(itemsets[s], transactions)
            if count/num_trans >= min_support:
                L.append(itemsets[s])
                supp_count.append(count)

            else:
                new_discarded.append(itemsets[s])
    return L, supp_count, new_discarded
            
        

In [18]:
def join_two_itemsets(it1,it2,columns):
    """
    Feature: join two itemsets
    """
    it1.sort(key=lambda x : columns.index(x))
    it2.sort(key=lambda x : columns.index(x))
    
    for i in range(len(it1)-1):
        if it1[i] != it2[i]:
            return []
    
    if columns.index(it1[-1]) < columns.index(it2[-1]):
        return it1 + [it2[-1]]
    return []

In [19]:
def join_set_itemsets(set_of_its, columns):
    """
    Feature: Join itemsets
    Return: items joined
    """
    C=[]
    for i in range(len(set_of_its)):
        for j in range(i+1,len(set_of_its)):
            it_out= join_two_itemsets(set_of_its[i],set_of_its[j],columns)
            if len(it_out)> 0:
                C.append(it_out)
    return C

In [20]:
from itertools import combinations,chain
def powerset(s):
    """
    This function returns a list containing all subsets of the input set, 
        including the empty set and the complete set.
    """
    return list(chain.from_iterable(combinations(s,r) for r in range(1,len(s)+1)))

In [21]:
def write_rules(X,X_S,S,conf,supp,lift,num_trans) :
    """
    Feaure: print association rules
    Agrument:
        X: Itemset in association rules
        X_S: Y in X->Y
        S: X in  X->Y
        conf: confidence
        supp:support
        lift:lift is an index used to measure the degree of association between 
             items in association rules
        num_trans:Count of transaction.
    """
    out_rules="\n \n"
    out_rules+= "Freq. Itemset: {}\n".format(X)
    out_rules+="   Rule:{} -> {}\n".format(list(S),list(X_S))
    out_rules+="   Conf: {0:2.3f}".format(conf)
    out_rules+="   Supp: {0:2.3f}".format(supp/num_trans)
    out_rules+="   Lift: {0:2.3f}".format(lift)
    return out_rules

In [22]:
def get_columns(transaction):
    """
    Returns elements contained in transactions
    """
    columns = set()
    for t in transaction:
        columns.update(set(t))
    columns = sorted(list(columns))
    return columns



### Hàm chính

In [23]:
def Apriori(transactions,min_sup, min_conf, min_lift=0):
        
    columns=get_columns(transaction)

    num_trans=len(transactions)
    C={}  # candidate itemset
    L={} # popular itemsets
    
    # Start k=1, create C1,L1
    itemset_size=1
    Discarded= {itemset_size:[]} # save element unpopular
    C.update({itemset_size:[ [f] for f in columns]})
    
    supp_count_L={} # frequency corresponding with L
    
    f,sup,new_discarded=get_frequent(C[itemset_size],transaction_new, min_sup, Discarded)
    
    Discarded.update({itemset_size :new_discarded})
    L.update({itemset_size: f})
    supp_count_L.update({itemset_size: sup})
    # End K=1
    
    
    # creat candidate itemset C[2...] and popular itemset L[2..]
    k=itemset_size+1 
    convergence= False
    while not convergence:
        C.update({k: join_set_itemsets(L[k-1],columns)})
        f,sup,new_discarded=get_frequent(C[k],transaction, min_sup, Discarded)
        Discarded.update({k: new_discarded})
        L.update({k: f})
        supp_count_L.update({k:sup})
        if len(L[k])==0:
            convergence=True
        k+= 1
        
    
    #Generate association rules
    assoc_rules_str=""
    for i in range(1,len(L)):
        for j in range(len(L[i])):
            s= list(powerset(set(L[i][j])))
            s.pop()
            for  z in s:
                S=set(z)
                X=set(L[i][j])
                X_S= set(X-S)
                sup_x=count_occurences(X,transaction)
                sup_x_s=count_occurences(X_S,transaction)
                conf=sup_x/count_occurences(S,transaction)
                lift=conf/(sup_x_s/num_trans)
                if conf >= min_conf and sup_x>= min_sup and lift>= min_lift:
                    assoc_rules_str+= write_rules(X,X_S,S,conf,sup_x,lift,num_trans)
    return assoc_rules_str
    
    

In [24]:
solution=Apriori(transaction_new,0.5,0.7)

In [25]:
print(solution)


 
Freq. Itemset: {'Bread', 'Apple'}
   Rule:['Bread'] -> ['Apple']
   Conf: 0.750   Supp: 0.545   Lift: 1.100
 
Freq. Itemset: {'Bread', 'Apple'}
   Rule:['Apple'] -> ['Bread']
   Conf: 0.800   Supp: 0.545   Lift: 1.100
 
Freq. Itemset: {'Butter', 'Apple'}
   Rule:['Butter'] -> ['Apple']
   Conf: 0.733   Supp: 0.500   Lift: 1.076
 
Freq. Itemset: {'Butter', 'Apple'}
   Rule:['Apple'] -> ['Butter']
   Conf: 0.733   Supp: 0.500   Lift: 1.076
 
Freq. Itemset: {'Milk', 'Apple'}
   Rule:['Apple'] -> ['Milk']
   Conf: 0.733   Supp: 0.500   Lift: 0.949
 
Freq. Itemset: {'Wine', 'Apple'}
   Rule:['Apple'] -> ['Wine']
   Conf: 0.733   Supp: 0.500   Lift: 1.008
 
Freq. Itemset: {'Bread', 'Butter'}
   Rule:['Bread'] -> ['Butter']
   Conf: 0.812   Supp: 0.591   Lift: 1.192
 
Freq. Itemset: {'Bread', 'Butter'}
   Rule:['Butter'] -> ['Bread']
   Conf: 0.867   Supp: 0.591   Lift: 1.192
 
Freq. Itemset: {'Bread', 'Milk'}
   Rule:['Bread'] -> ['Milk']
   Conf: 0.812   Supp: 0.591   Lift: 1.051
 
Freq.

## So sánh hàm có sẵn và tự viết

### So sánh khi chỉ có hai tham số là min_sup và min_cof

#### Hàm có sẵn

In [26]:
assocciation_rules=apriori(transaction_new,min_support=0.5, min_confidence=0.7)
assocciation_rules=list(assocciation_rules)
output_solution(assocciation_rules)

                     Rule   Support Confidence      Lift
0                => Bread  0.727273   0.727273       1.0
1                 => Milk  0.772727   0.772727       1.0
2                 => Wine  0.727273   0.727273       1.0
3          Apple => Bread  0.545455        0.8       1.1
4          Bread => Apple  0.545455       0.75       1.1
5         Apple => Butter       0.5   0.733333  1.075556
6         Butter => Apple       0.5   0.733333  1.075556
7           Apple => Milk       0.5   0.733333   0.94902
8           Apple => Wine       0.5   0.733333  1.008333
9         Bread => Butter  0.590909     0.8125  1.191667
10        Butter => Bread  0.590909   0.866667  1.191667
11          Bread => Milk  0.590909     0.8125  1.051471
12          Milk => Bread  0.590909   0.764706  1.051471
13          Bread => Wine  0.590909     0.8125  1.117188
14          Wine => Bread  0.590909     0.8125  1.117188
15         Butter => Milk  0.590909   0.866667  1.121569
16         Milk => Butter  0.59

#### Hàm viết tay

In [27]:
print(Apriori(transaction_new,0.5,0.7))


 
Freq. Itemset: {'Bread', 'Apple'}
   Rule:['Bread'] -> ['Apple']
   Conf: 0.750   Supp: 0.545   Lift: 1.100
 
Freq. Itemset: {'Bread', 'Apple'}
   Rule:['Apple'] -> ['Bread']
   Conf: 0.800   Supp: 0.545   Lift: 1.100
 
Freq. Itemset: {'Butter', 'Apple'}
   Rule:['Butter'] -> ['Apple']
   Conf: 0.733   Supp: 0.500   Lift: 1.076
 
Freq. Itemset: {'Butter', 'Apple'}
   Rule:['Apple'] -> ['Butter']
   Conf: 0.733   Supp: 0.500   Lift: 1.076
 
Freq. Itemset: {'Milk', 'Apple'}
   Rule:['Apple'] -> ['Milk']
   Conf: 0.733   Supp: 0.500   Lift: 0.949
 
Freq. Itemset: {'Wine', 'Apple'}
   Rule:['Apple'] -> ['Wine']
   Conf: 0.733   Supp: 0.500   Lift: 1.008
 
Freq. Itemset: {'Bread', 'Butter'}
   Rule:['Bread'] -> ['Butter']
   Conf: 0.812   Supp: 0.591   Lift: 1.192
 
Freq. Itemset: {'Bread', 'Butter'}
   Rule:['Butter'] -> ['Bread']
   Conf: 0.867   Supp: 0.591   Lift: 1.192
 
Freq. Itemset: {'Bread', 'Milk'}
   Rule:['Bread'] -> ['Milk']
   Conf: 0.812   Supp: 0.591   Lift: 1.051
 
Freq.

### So sánh với ba tham số min_sup, min_conf, min_lift

#### Hàm có sẵn

In [28]:
assocciation_rules=apriori(transaction_new,min_support=0.5, min_confidence=0.7, min_lift=1.2)
assocciation_rules=list(assocciation_rules)
output_solution(assocciation_rules)

                    Rule Support Confidence      Lift
0  Butter => Bread, Milk     0.5   0.733333  1.241026
1  Bread, Milk => Butter     0.5   0.846154  1.241026


#### Hàm viết tay

In [29]:
print(Apriori(transaction_new,0.5,0.7,1.2))


 
Freq. Itemset: {'Bread', 'Milk', 'Butter'}
   Rule:['Butter'] -> ['Bread', 'Milk']
   Conf: 0.733   Supp: 0.500   Lift: 1.241
 
Freq. Itemset: {'Bread', 'Milk', 'Butter'}
   Rule:['Bread', 'Milk'] -> ['Butter']
   Conf: 0.846   Supp: 0.500   Lift: 1.241
